extensions/interactions/GraphInput/directives/graph-viz.component.ts

```diff
@@ -0,0 +1,626 @@
+        x: this.state.mouseX,
+        y: this.state.mouseY,
+        label: ''
+      });
+    }
+    if (this.state.hoveredVertex === null) {
+      this.state.selectedVertex = null;
+    }
+    if (this.state.hoveredEdge === null) {
+      this.state.selectedEdge = null;
+    }
+  }
+
+  initButtons(): void {
+    this.buttons = [];
+    if (this.canMoveVertex) {
+      this.buttons.push({
+        text: '\uF0B2',
+        description: 'I18N_INTERACTIONS_GRAPH_MOVE',
+        mode: this._MODES.MOVE
+      });
+    }
+    if (this.canAddEdge) {
+      this.buttons.push({
+        text: '\uF0C1',
+        description: 'I18N_INTERACTIONS_GRAPH_ADD_EDGE',
+        mode: this._MODES.ADD_EDGE
+      });
+    }
+    if (this.canAddVertex) {
+      this.buttons.push({
+        text: '\uF067',
+        description: 'I18N_INTERACTIONS_GRAPH_ADD_NODE',
+        mode: this._MODES.ADD_VERTEX
+      });
+    }
+    if (this.canDeleteVertex || this.canDeleteEdge) {
+      this.buttons.push({
+        text: '\uF068',
+        description: 'I18N_INTERACTIONS_GRAPH_DELETE',
+        mode: this._MODES.DELETE
+      });
+    }
+  }
+
+  init(): void {
+    this.initButtons();
+    this.state.currentMode = this.buttons[0].mode;
+    if (this.isMobile) {
+      if (this.state.currentMode === this._MODES.ADD_EDGE) {
+        this.helpText =
+          'I18N_INTERACTIONS_GRAPH_EDGE_INITIAL_HELPTEXT';
+      } else if (this.state.currentMode === this._MODES.MOVE) {
+        this.helpText =
+          'I18N_INTERACTIONS_GRAPH_MOVE_INITIAL_HELPTEXT';
+      } else {
+        this.helpText = '';
+      }
+    } else {
+      this.helpText = '';
+    }
+  }
+
+  toggleGraphOption(option: string):void {
+    // Handle the case when we have two edges s -> d and d -> s.
+    if (option === 'isDirected' && this.graph[option]) {
+      this._deleteRepeatedUndirectedEdges();
+    }
+    this.graph[option] = !this.graph[option];
+  }
+
+  setMode(mode: number): void {
+    this.state.currentMode = mode;
+    if (this.isMobile) {
+      if (this.state.currentMode === this._MODES.ADD_EDGE) {
+        this.helpText =
+          'I18N_INTERACTIONS_GRAPH_EDGE_INITIAL_HELPTEXT';
+      } else if (this.state.currentMode === this._MODES.MOVE) {
+        this.helpText =
+          'I18N_INTERACTIONS_GRAPH_MOVE_INITIAL_HELPTEXT';
+      } else {
+        this.helpText = null;
+      }
+    } else {
+      this.helpText = null;
+    }
+    this.state.addEdgeVertex = null;
+    this.state.selectedVertex = null;
+    this.state.selectedEdge = null;
+    this.state.currentlyDraggedVertex = null;
+    this.state.hoveredVertex = null;
+  }
+
+  onClickModeButton(mode: number, event: Event): void {
+    event.preventDefault();
+    event.stopPropagation();
+    if (this.interactionIsActive) {
+      this.setMode(mode);
+    }
+  }
+
+  // TODO(czx): Consider if there's a neat way to write a reset()
+  // function to clear bits of this.state
+  // (e.g. currentlyDraggedVertex, addEdgeVertex).
```

Comment:
Again, maybe replace this with an issue and use the issue number here.

Reply:
https://github.com/oppia/oppia/pull/12090#discussion_r586676996